### PR TITLE
Honor Context Deadlines in CallOptionsFactory.

### PR DESCRIPTION
**Pull request by sduskis**

### CODE REVIEW

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/CallOptionsFactory.java [hunk 1]
```diff
@@ -58,7 +59,12 @@
     @Override
     public <RequestT> CallOptions create(MethodDescriptor<RequestT, ?> descriptor,
         RequestT request) {
-      return CallOptions.DEFAULT;
+      Deadline contextDeadline = Context.current().getDeadline();
```
Comment (kevinsi4508) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/CallOptionsFactory.java]: Should we make Default CallOptionsFactory just return CallOptions.DEFAULT, like what the comment says?

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/CallOptionsFactory.java [hunk 2]
```diff
@@ -71,17 +77,28 @@ public ConfiguredCallOptionsFactory(CallOptionsConfig config) {
...
+     * Creates a {@link CallOptions} with a focus on {@link Deadlines}.  Deadlines are decided in the following order:
+     * <ol>
+     *     <li> If a user set a  {@link Context} deadline (see {@link Context#getDeadline()}), use that</li>
+     *     <li> If a user configured deadlines via {@link CallOptionsConfig}, use it.</li>
+     *     <li> Otherwise, use {@link CallOptions#DEFAULT}.</li>
+     * </ol>
+     *
+     */
+    public <RequestT> CallOptions create(MethodDescriptor<RequestT, ?> descriptor, RequestT request) {
+      Deadline contextDeadline = Context.current().getDeadline();
```
Comment (kevinsi4508) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/CallOptionsFactory.java]: How does the deadline in Context set by user?
It seems to be harder to track where the deadline is coming from.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/CallOptionsFactory.java (hunk 1 above)
Comment (sduskis) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/CallOptionsFactory.java]: I'll update the comments.  We should honor the context deadline.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/CallOptionsFactory.java (hunk 2 above)
Comment (sduskis) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/CallOptionsFactory.java]: See the test for how a user would set the context deadline.  Where the operation deadline comes from is kind of irrelevant.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/CallOptionsFactory.java [hunk 3]
```diff
@@ -71,17 +77,28 @@ public ConfiguredCallOptionsFactory(CallOptionsConfig config) {
...
+      if (contextDeadline != null) {
+        return CallOptions.DEFAULT.withDeadline(contextDeadline);
+      } else if (!config.isUseTimeout() || request == null) {
```
Comment (kevinsi4508) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/CallOptionsFactory.java]: nit: Can we have the same order as in the comment?
- use deadline from context
- use deadline from config if provided
- use DEFAULT

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AbstractRetryingOperation.java [hunk 4]
```diff
@@ -26,7 +26,6 @@
 
 import com.google.api.client.util.BackOff;
 import com.google.api.client.util.Sleeper;
-import com.google.bigtable.v2.BigtableGrpc;
```
Comment (kevinsi4508) [inline bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AbstractRetryingOperation.java]: Separate PR for clean up?

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    public void testDefaultWithContext() {
+        Context.CancellableContext context = Context.current().withDeadlineAfter(1, TimeUnit.SECONDS, mockExecutor);
+        context.run(new Runnable() {
+            @Override
+            public void run() {
+                CallOptionsFactory factory = new CallOptionsFactory.Default();
+                Assert.assertEquals(
+                        Context.current().getDeadline().timeRemaining(TimeUnit.MILLISECONDS),
+                        getDeadlineMs(factory, null),
+                        100.0
```
Comment (kevinsi4508) [inline bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java]: Cast to int instead? 100.0 delta seems big.

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        CallOptionsFactory factory = new CallOptionsFactory.ConfiguredCallOptionsFactory(config);
+        Assert.assertSame(CallOptions.DEFAULT, factory.create(null, null));
+    }
+
+    @Test
+    public void testConfiguredConfigEnabled() {
+        CallOptionsConfig config = new CallOptionsConfig.Builder().setUseTimeout(true).build();
+        CallOptionsFactory factory = new CallOptionsFactory.ConfiguredCallOptionsFactory(config);
+        Assert.assertEquals(
+                (double) config.getShortRpcTimeoutMs(),
```
Comment (kevinsi4508) [inline bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java]: Could you explicitly set ShortRpcTimeoutMs? It is easier to understand where this value is coming from.

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        Assert.assertEquals(
+                (double) config.getLongRpcTimeoutMs(),
+                (double) getDeadlineMs(factory, MutateRowsRequest.getDefaultInstance()),
+                100.0
+        );
+    }
+
+    @Test
+    public void testConfiguredWithContext() {
+        Context.CancellableContext context = Context.current().withDeadlineAfter(1, TimeUnit.SECONDS, mockExecutor);
```
Comment (kevinsi4508) [inline bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java]: Keep the config useTimeout == true? then we know it takes precedence over config.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/CallOptionsFactory.java (hunk 1 above)
Comment (sduskis) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/CallOptionsFactory.java]: I've updated the comments.

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java (hunk 5 above)
Comment (sduskis) [reply bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java]: What I really want is a predictable Deadline.  I'll change the test to use [this method](https://github.com/grpc/grpc-java/blob/master/context/src/main/java/io/grpc/Deadline.java#L51], which is a bit hacky.

File: bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AbstractRetryingOperation.java (hunk 4 above)
Comment (kevinsi4508) [reply bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AbstractRetryingOperation.java]: Sure.

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/io/grpc/DeadlineUtil.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+ */
+package io.grpc;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Hack to allow the creation of a {@link Deadline} with a constant time.
+ */
+public class DeadlineUtil {
+  static Deadline.Ticker constantBasedTicker(final long timeNs) {
```
Comment (kevinsi4508) [inline bigtable-client-core-parent/bigtable-client-core/src/test/java/io/grpc/DeadlineUtil.java]: nit: nowNanoSeconds or nowNanos or nowNs or now?
I feel "now" is better reflect the fact that the time is fixed.

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/io/grpc/DeadlineUtil.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    return new Deadline.Ticker() {
+      @Override
+      public long read() {
+        return timeNs;
+      }
+    };
+  }
+
+  public static Deadline deadlineWithFixedTime(int duration, TimeUnit unit, long timeNs) {
+    return Deadline.after(duration, unit, constantBasedTicker(timeNs));
```
Comment (kevinsi4508) [inline bigtable-client-core-parent/bigtable-client-core/src/test/java/io/grpc/DeadlineUtil.java]: Weird, there is no Duration that contains both duration value and its unit?

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+  @Test
+  public void testDefaultWithContext() {
+    Deadline deadline = DeadlineUtil.deadlineWithFixedTime(1, TimeUnit.SECONDS, NOW);
+    Context.CancellableContext context = Context.current().withDeadline(deadline, mockExecutor);
+    context.run(new Runnable() {
+      @Override
+      public void run() {
+        CallOptionsFactory factory = new CallOptionsFactory.Default();
+        assertEqualsDeadlines(
+                Context.current().getDeadline().timeRemaining(TimeUnit.MILLISECONDS),
```
Comment (kevinsi4508) [inline bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java]: deadline.timeRemain(...) ?

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java [hunk 11]
```diff
@@ -31,81 +33,92 @@
...
-    }
+  private static void assertSameTimeDiff(long l1, long l2) {
+    Assert.assertEquals((double) l1, (double) l2, 10);
+  }
 
-    private double getDeadlineMs(CallOptionsFactory factory, Object request) {
-        return (double) factory.create(null, request).getDeadline().timeRemaining(TimeUnit.MILLISECONDS);
-    }
+  private int getDeadlineMs(CallOptionsFactory factory, Object request) {
+    return (int) factory.create(null, request).getDeadline().timeRemaining(TimeUnit.MILLISECONDS);
```
Comment (kevinsi4508) [inline bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java]: Do we need to cast to int here? Maybe just use long?

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java [hunk 12]
```diff
@@ -31,81 +33,92 @@
...
 
-    @Test
-    public void testConfiguredDefaultConfig() {
-        CallOptionsConfig config = new CallOptionsConfig.Builder().build();
-        CallOptionsFactory factory = new CallOptionsFactory.ConfiguredCallOptionsFactory(config);
-        Assert.assertSame(CallOptions.DEFAULT, factory.create(null, null));
-    }
+  @Test
+  public void testConfiguredConfigEnabled() {
+    CallOptionsConfig config = new CallOptionsConfig.Builder().setUseTimeout(true).build();
```
Comment (kevinsi4508) [inline bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/TestCallOptionsFactory.java]: Not sure if you have addressed my previous comment.
Set ShortRpcTimeoutMs implicitly? Easier to see where it is coming from.

**Comment by kevinsi4508** [comment]
Few minor things, otherwise LGTM

File: bigtable-client-core-parent/bigtable-client-core/src/test/java/io/grpc/DeadlineUtil.java (hunk 9 above)
Comment (sduskis) [reply bigtable-client-core-parent/bigtable-client-core/src/test/java/io/grpc/DeadlineUtil.java]: I'm mimicking the grpc `Deadline` API.  It doesn't use a `Duration` object.

**Pull request merged**